Add data generation code and Update lodbrok

me-diru opened the pull request:
- Added dummy data generation code with relevant doc files in the data folder.
- Trained and evaluated lodbrok with the latest dummy data, documented the same with results.
- Organized the files in proper folders and deleted unnecessary files.

me-diru committed:
Add LodBrok explanation doc

Documented all info related to LodBrok done till now.
Updated README.md with link to LodBrok.md.
Ran lodbrok_evaluation.ipynb locally.

me-diru committed:
Add lodbrok notebook to show LodBrok internals

Added lobrok.ipynb file to show how the model works with
supportive documentation. Also updated link in README.md.

me-diru committed:
Add requirements.txt to run LodBrok

Contains necessary packages needed to run LodBrok

me-diru committed:
Add eval.py and Update lodbrok.py

eval.py is used to evaluate LodBrok with dummy data for testing
purposes and lodbrok.py is upadated to use keras API.

me-diru committed:
Add dataset to train LodBrok

me-diru committed:
Switch from unused lodbrok.ipynb to LodBrok.md

Changes in LodBrok.md:
    - Added requirements needed to run lodbrok
    - Updated train and evaluate info with instructions and images.
Removed unused lodbrok.ipynb

me-diru committed:
Add spam dummy data generator code and it's dependencies

spam_dummy_data_generator.ipynb is used to generate spam data required
to train LodBrok.
domains_data.csv file is used to extract spam domains.
requirements_dummy_data has necessary packages to run notebook.
Create DummyData.md with info on why/how to use notebook.

me-diru committed:
Add non_spam generator code and it's dependencies

Added non_spam generator.ipynb to create non_spam data.
Updated domains_data.csv and DummyData.md to include non_spam info.
Corrected errors in spam_generator.ipynb.

me-diru committed:
Update data generation notebooks

Updated dataset_generation notebook by:
   - Changed input files for generated dummy data
   - Changed a typo in preprocess_editor funciton
   - Kept in data folder
Changed size of dummy data generated

me-diru committed:
Update and Organize LodBrok trained with dummy data

- Organized all the files properly in each seperate folder.
- Updated the LodBrok code to train on dummy data.
- Changed the summary.png and eval.png with latest results.
- Rename DummyData.md to README.md for better visibility.

LeoVerto commented on data/dataset_generation.ipynb:
```diff
@@ -189,7 +197,7 @@
     "    domain = urlparse(editor[\"website\"]).hostname\n",
     "    if domain is not None:\n",
-    "        website_domain = email_tokenizer.texts_to_sequences([editor[\"email\"].split(\"@\")[1]])[0]\n",
+    "        website_domain = website_tokenizer.texts_to_sequences(urlparse(editor[\"website\"]).hostname)[0]\n",
```
Good job catching this!

LeoVerto commented on models/lodbrok.py:
```diff
@@ -11,8 +11,8 @@ def get_model() -> Model:
 
-    email_embedding = Embedding(output_dim=256, input_dim=1024, input_length=1, name="email_embedding")(email_input)
-    website_embedding = Embedding(output_dim=256, input_dim=1024, input_length=1, name="website_embedding")(website_input)
+    email_embedding = Embedding(output_dim=256, input_dim=1025, input_length=1, name="email_embedding")(email_input)
+    website_embedding = Embedding(output_dim=256, input_dim=1025, input_length=1, name="website_embedding")(website_input)
```
Why did change the input dimensions here?

me-diru replied:
I was getting  [this error](https://github.com/tensorflow/tensorflow/issues/23698) with existing dimensions.

LeoVerto reviewed (commented):
I really like how you centralized all the documentation and documented all the missing parts!

The dummy data generation code seems sensible and like something really useful for testing new models.

Do you think committing the model snapshots to the repo makes sense? In general I'd rather be cautious about committing large binary files to a repo, after all you'll never get them out of the git history.

Other than that I've left a couple comments but in general this PR looks good to me!

me-diru commented:
> I really like how you centralized all the documentation and documented all the missing parts!

Yea, @yvanzo helped me structure the repo and make a quality PR :) 

>  The dummy data generation code seems sensible and like something really useful for testing new models.


Yup, I used your analysis and the spam_data that I had to come up with that logic.

> Do you think committing the model snapshots to the repo makes sense? In general I'd rather be cautious about committing large binary files to a repo, after all you'll never get them out of the git history.

Yes, you are right. Model snapshots aren't necessary.

> Other than that I've left a couple comments but in general this PR looks good to me!

Answered, thanks for a thorough review.

yvanzo reviewed (changes requested):
The large files that have been removed should not be committed at all, not even in intermediate commits.
Since they are no more in the working tree, there should be an automated way to retrieve them.

me-diru commented:
> The large files that have been removed should not be committed at all, not even in intermediate commits.
> Since they are no more in the working tree, there should be an automated way to retrieve them.

I have removed all pickle files in GitHub history which were unnecessary and also renamed my data related commit message to show the same.
 
P.S: I followed [this process](https://stackoverflow.com/a/43640996/9824544) to clean history.

me-diru committed:
Add script to download binaries

- Used to download dataset and tokenizer files.

yvanzo commented on download_script.py:
```diff
@@ -0,0 +1,50 @@
+import requests
+import argparse
+
+# help function for better understanding of the scripts functionality
+
```
No longer needed since it now precedes the rest of the script, just remove it.

```suggestion
```

yvanzo commented on download_script.py:
```diff
@@ -0,0 +1,50 @@
+
+# help function for better understanding of the scripts functionality
+
+description = """  This script is used to download all the pickle/dataset files needed to train, evaluate and predict using the model.
```
That seems to be a typo as both tokenizer and dataset files are pickles:
```suggestion
description = """  This script is used to download all the tokenizer/dataset pickle files needed to train, evaluate and predict using the model.
```

By the way, this script should rather be named `get_pickles.py` which would be more self-explanatory.

yvanzo commented on download_script.py:
```diff
@@ -0,0 +1,50 @@
+# help function for better understanding of the scripts functionality
+
+description = """  This script is used to download all the pickle/dataset files needed to train, evaluate and predict using the model.
+                    Also downloads model tokenizers used for preprocessing the datasets before feeding them to the model \n
+
+                   The script assumes to be run at spambrainz_ml top directory \n
+
+                   The files if present are overwritten by the script, kindly be careful"""
```
1. Full stops `.``are missing at the end of each sentence.
2. `\n`s and line breaks are ignored by the default formatter. Either remove them or use another [formatter](https://docs.python.org/3/library/argparse.html#formatter-class) such as `RawDescriptionHelpFormatter`.

yvanzo commented on download_script.py:
```diff
@@ -0,0 +1,50 @@
+email_tokenizer_request = requests.get(email_tokenizer_url, allow_redirects=True)
+open('data/email_tokenizer.pickle', 'wb').write(bio_tokenizer_request.content)
+
+# dataset files to for model training, evaluation and predictions
```
```suggestion
# Dataset files to for model training, evaluation and predictions
```

yvanzo commented on download_script.py:
```diff
@@ -0,0 +1,50 @@
+bio_tokenizer_url = 'https://github.com/metabrainz/spambrainz_ml/releases/download/v-0.1/bio_tokenizer.pickle'
+web_tokenizer_url = 'https://github.com/metabrainz/spambrainz_ml/releases/download/v-0.1/web_tokenizer.pickle'
+email_tokenizer_url = 'https://github.com/metabrainz/spambrainz_ml/releases/download/v-0.1/email_tokenizer.pickle'
+
+
```
Please drop this extraneous blank line and trailing blank lines at the end of the script.

```suggestion

```

yvanzo commented on download_script.py:
```diff
@@ -0,0 +1,50 @@
+import requests
+import argparse
```
Order suggested by `pylint` among other things:

```suggestion
import argparse
import requests
```

yvanzo reviewed (approved):
Thanks!
